### PR TITLE
go.mod: update cloud.google.com/go/storage to v1.16.0

### PR DESCRIPTION
**Pull request by adityamaru**
https://github.com/googleapis/google-cloud-go/releases/tag/storage%2Fv1.16.0 has the sought-after GOAWAY fix and is the latest release of the storage subpackage.

Release note: None

### CODE REVIEW

**Comment by adityamaru** [comment]
The PR is likely going to run into `make bazel_generate` failing with:

```
ERROR: /private/var/tmp/_bazel_adityamaru/2592a7600a28e43d0b132c217219adf9/external/io_bazel_rules_go/BUILD.bazel:73:16: in go_context_data rule @io_bazel_rules_go//:go_context_data: cycle in dependency graph:
    //:gazelle
    //:gazelle-runner
    @bazel_gazelle//cmd/gazelle:gazelle
    @bazel_gazelle//language/go:go_default_library
    @bazel_gazelle//language/go:go
    @com_github_pelletier_go_toml//:go-toml
.-> @io_bazel_rules_go//:go_context_data
|   @io_bazel_rules_nogo//:nogo
|   @io_bazel_rules_go//:default_nogo
|   @io_bazel_rules_go//go/tools/builders:nogo_srcs
|   @org_golang_x_tools//go/analysis:go_tool_library
|   @org_golang_x_tools//internal/analysisinternal:go_tool_library
|   @org_golang_x_tools//go/ast/astutil:go_tool_library
|   @org_golang_x_tools//internal/typeparams:typeparams
`-- @io_bazel_rules_go//:go_context_data
ERROR: Analysis of target '//:gazelle' failed; build aborted
```

This looks similar to what https://github.com/cockroachdb/cockroach/pull/66880 is blocked on which is being tracked at https://github.com/bazelbuild/rules_go/issues/2479.

**Review by pbardea — APPROVED** [review]
Changes look good (and took a look at the release notes); LGTM pending resolution of the Bazel blocker.

**Comment by adityamaru** [comment]
Replaced by https://github.com/cockroachdb/cockroach/pull/67806

**Pull request closed without merging**
